lib/SimpleSAML/Metadata/MetaDataStorageHandlerMDX.php

```diff
@@ -0,0 +1,292 @@
+		}
+
+		if (!is_array($data)) {
+			throw new Exception('Cached metadata from "' . $cachefilename . '" wasn\'t an array.');
+		}
+
+		return $data;
+	}
+
+
+	/**
+	 * Save a entity to the cache.
+	 *
+	 * @param string $set  The metadata set this entity belongs to.
+	 * @param string $entityId  The entity id of this entity.
+	 * @param array $data  The associative array with the metadata for this entity.
+	 */
+	private function writeToCache($set, $entityId, $data) {
+		assert('is_string($set)');
+		assert('is_string($entityId)');
+		assert('is_array($data)');
+
+		if (empty($this->cacheDir)) {
+			return;
+		}
+
+		$cachefilename = $this->getCacheFilename($set, $entityId);
+		if (!is_writable(dirname($cachefilename))) throw new Exception('Could not write cache file for entity [' . $cachefilename. ']');
+		SimpleSAML_Logger::debug('MetaData - Handler.MDX: Writing cache [' . $entityId . '] => [' . $cachefilename . ']' );
+		file_put_contents($cachefilename, serialize($data));
+	}
+
+
+	/**
+	 * Retrieve metadata for the correct set from a SAML2Parser.
+	 *
+	 * @param SimpleSAML_Metadata_SAMLParser $entity  A SAML2Parser representing an entity.
+	 * @param string $set  The metadata set we are looking for.
+	 * @return array|NULL  The associative array with the metadata, or NULL if no metadata for
+	 *                     the given set was found.
+	 */
+	private static function getParsedSet(SimpleSAML_Metadata_SAMLParser $entity, $set) {
+		assert('is_string($set)');
+
+		switch($set) {
+		case 'saml20-idp-remote':
+			return $entity->getMetadata20IdP();
+		case 'saml20-sp-remote':
+			return $entity->getMetadata20SP();
+		case 'shib13-idp-remote':
+			return $entity->getMetadata1xIdP();
+		case 'shib13-sp-remote':
+			return $entity->getMetadata1xSP();
+		case 'attributeauthority-remote':
+			return $entity->getAttributeAuthorities();
+
+		default:
+			SimpleSAML_Logger::warning('MetaData - Handler.MDX: Unknown metadata set: ' . $set);
+		}
+
+		return NULL;
+	}
+
+
+	/**
+	 * Overriding this function from the superclass SimpleSAML_Metadata_MetaDataStorageSource.
+	 *
+	 * This function retrieves metadata for the given entity id in the given set of metadata.
+	 * It will return NULL if it is unable to locate the metadata.
+	 *
+	 * This class implements this function using the getMetadataSet-function. A subclass should
+	 * override this function if it doesn't implement the getMetadataSet function, or if the
+	 * implementation of getMetadataSet is slow.
+	 *
+	 * @param $index  The entityId or metaindex we are looking up.
+	 * @param $set  The set we are looking for metadata in.
+	 * @return An associative array with metadata for the given entity, or NULL if we are unable to
+	 *         locate the entity.
+	 */
+	public function getMetaData($index, $set) {
+		assert('is_string($index)');
+		assert('is_string($set)');
+
+		if (!preg_match('@(https?://([-\w\.]+)+(:\d+)?(/([\w/_\.]*(\?\S+)?)?)?)@', $index)) {
```

Comment:
Why are you enforcing the entity ID to be a URL? The SAML spec doesn't mandate it to be a URL, so it could virtually be anything. In fact it's quite common to have URNs as entity IDs...